[plugin/hw] Add more HW accelerator and update implementation

kparichay opened the pull request:
Add TPU as a backend accelerator property option in list of hardware
accelerators
Move default accelerator to avoid changing default accelerator enum value
everytime a new backend device is added

Refer #1849 

Self evaluation:
Build test: [x]Passed [ ]Failed [ ]Skipped
Run test: [x]Passed [ ]Failed [ ]Skipped

Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

dongju-chae commented on gst/nnstreamer/nnstreamer_plugin_api_filter.h:
```diff
@@ -28,23 +28,25 @@
 #define REGEX_ACCL_EACH_ELEM(ELEM) "[^!]" ELEM "[,]?"
 
 #define REGEX_ACCL_AUTO REGEX_ACCL_EACH_ELEM(ACCL_AUTO_STR)
+#define REGEX_ACCL_DEF  REGEX_ACCL_EACH_ELEM(ACCL_DEF_STR)
```
Minor comment: it's better to move this line above `ACCL_AUTO_STR`?

kparichay replied:
Done

dongju-chae commented on gst/nnstreamer/nnstreamer_plugin_api_filter.h:
```diff
@@ -76,18 +78,19 @@ extern "C" {
 
   /** If there is no default config, and device needs to be specified, fallback to ACCL_AUTO */
-  ACCL_DEFAULT = 0x20000    /**< use default device configuration by the framework*/
+  ACCL_DEFAULT = 0x1,     /**< use default device configuration by the framework*/
```
need a space `framework` and `*/`

kparichay replied:
Done

myungjoo commented on gst/nnstreamer/nnstreamer_plugin_api_filter.h:
```diff
@@ -28,23 +28,25 @@
 #define REGEX_ACCL_NEON REGEX_ACCL_EACH_ELEM(ACCL_NEON_STR)
 #define REGEX_ACCL_SRCN REGEX_ACCL_EACH_ELEM(ACCL_SRCN_STR)
-#define REGEX_ACCL_DEF  REGEX_ACCL_EACH_ELEM(ACCL_DEF_STR)
+#define REGEX_ACCL_TPU  REGEX_ACCL_EACH_ELEM(ACCL_TPU_STR)
```
We need to be more specific if we are going to designate an enum for each model of accelerator.

kparichay replied:
Currently, these enums are generic ones. For example - ACCL_NPU denotes a generic NPU device. Which device will be used if chosen this accelerator (if multiple accelerators present) is dependent on the framework.
In the frameworks supported as extensions of tensor filter with nnstreamer, they do not give the option to choose the accelerator specific model out of the available ones.

kparichay replied:
Embedding model specific information in these enums might be too much.
How about - model specific information can be taken in PROP_ACCELERATOR as input, parsed to be stored separately, and can be used later when frameworks support it?

myungjoo replied:
(This is suggestion, we'll need a lot of revises)
Keeping compatible with ml_nnfw_hw_e of nnstreamer.h, we may have parsing rules as:
```
cpu
gpu
npu (== Any NPU, 0x3000)
npu.sr (== Any SR NPU 0x3800)
npu.srasr (== SR-ASR-NPU 0x3801)
npu.srvision (== SR-Vision-NPU 0x3082)
npu.movidius (0x3001)
npu.edgetpu (== Google/Coral TPU, 0x3002)
npu.vivante (== Vivante NPU. 0x3003)
...
??

myungjoo replied:
And we need some explanations in the code; e.g., what is "SRCN"?

kparichay replied:
Good suggestion - updated
Current implementation is not compatible with ml_nnfw_hw_e - 
NPU enum value is 3000 in ml_nnfw_hw_e and is 4000 internally (to avoid bit overlap) - should the enum value be updated to 3000 internally as well?

kparichay commented:
V2:
- Updated the regex to be formed at run-time using a list of supported accelerators.
- Accelerators are also stored as GEnumValues to do conversion from string to int value and vice versa easily
- More accelerators are added in heirarchical order. srcn and neon moved to npu and cpu heirarchy correspondingly.
- Added more accelerators to be supported in future.

@dongju-chae Please review once more due to the significance of changes added in V2

jaeyun-jung commented on ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite.cc:
```diff
@@ -520,6 +510,7 @@ TFLiteCore::TFLiteCore (const char * _model_path, const char * accelerators)
 
   setAccelerator (accelerators);
 
+  g_message ("nnapi = %d, accl = %s", use_nnapi, get_accl_hw_str(accelerator));
```
this line is duplicated with below line (DBG)

kparichay replied:
Updated.

jaeyun-jung commented on gst/nnstreamer/tensor_filter/tensor_filter_common.c:
```diff
@@ -70,6 +89,68 @@ enum
+ * @brief to get and register hardware accelerator backend enum
+ */
+static GType
+accl_hw_get_type (void)
```
recommended to move original code here,  and fix indent in line 105

kparichay replied:
Moving the original code here does allow to use the G_GNUC_CONST attribute as this attribute can only be set in function declaration.

kparichay replied:
Line 105 current indentation is the needed indentation to pass indent CI test.
```
static gchar *
strcpy2 ()
```
Updating to above fails CI indent test.

jaeyun-jung commented on gst/nnstreamer/nnstreamer_plugin_api_filter.h:
```diff
@@ -252,8 +253,8 @@ get_accl_hw_str (const accl_hw key);
 extern accl_hw
-parse_accl_hw (const char * accelerators, const char * regex_accl,
-    const char * regex_accl_elem);
+parse_accl_hw (const char * accelerators,
```
recommended to define prototype in one line.

kparichay replied:
Updated.
